File: 2.Add-Two-Numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    int offset = 0;
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if ((l1 != null) && (l2 != null)) {
            int sum = l1.val + l2.val + this.offset;
            if (sum < 10)
                this.offset = 0;
            else {
                sum %= 10;
                this.offset = 1;
            }
            return new ListNode(sum, addTwoNumbers(l1.next, l2.next));
        }
        if (l1 != null) {
            int sum = l1.val + this.offset;
            if (sum < 10)
                this.offset = 0;
            else {
                this.offset = 1;
                sum %= 10;
            }
            return new ListNode(sum, addTwoNumbers(l1.next, null));
        }
        if (l2 != null) {
            int sum = l2.val + this.offset;
            if (sum < 10)
                this.offset = 0;
            else {
                this.offset = 1;
                sum %= 10;
            }
            return new ListNode(sum, addTwoNumbers(l2.next, null));
        }
        if (this.offset == 1)
            return new ListNode(1);
        return null;
    }
}

File: 2.Add-Two-Numbers/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
private:
    int offset = 0;
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int sum;
        if (l1 != nullptr && l2 != nullptr) {
            sum = l1->val + l2->val + offset;
            offset = sum / 10;
            return new ListNode(sum % 10, addTwoNumbers(l1->next, l2->next));
        }
        ListNode *ptr = l1 ? l1 : l2;
        ListNode result;
        ListNode *ptrCurr = &result;
        while (ptr) {
            sum = ptr->val + offset;
            offset = sum / 10;
            ptrCurr->next = new ListNode(sum % 10);
            ptrCurr = ptrCurr->next;
            ptr = ptr->next;
        }
        if (offset == 1)
            ptrCurr->next = new ListNode(offset);
        return result.next;        
    }
};